model/dns_zone/dns_zone.rb

```diff
@@ -31,5 +37,50 @@ def delete_record(record_name:, type: nil, data: nil)
     records.each do |r|
       r.destroy
     end
+
+    command = "zone-unset #{name} #{record_name} #{type} #{data}"
+    run_on_all_name_servers(command, "no such node in zone found")
+  end
+
+  def run_on_all_name_servers(command, ignore)
+    lock_name_servers do
+      name_servers.flat_map(&:vms).each do |vm|
+        # knotc also supports passing multiple commands via stdin, but we
+        # don't use that, because exception handling would be much more
+        # difficult with that.
+        begin
+          vm.sshable.cmd("sudo -u knot knotc zone-abort #{name}")
+        rescue Sshable::SshError => ex
+          raise unless ex.stdout.include?("no active transaction")
+        end
+
+        vm.sshable.cmd("sudo -u knot knotc zone-begin #{name}")
+
+        begin
+          vm.sshable.cmd("sudo -u knot knotc #{command}")
+        rescue Sshable::SshError => ex
+          raise unless ex.stdout.include?(ignore)
+        end
+
+        vm.sshable.cmd("sudo -u knot knotc zone-commit #{name}")
+      end
+    end
+  end
+
+  def lock_name_servers
+    DB.transaction do
+      # We take an advisory lock for each name server to ensure only one DNS
+      # operation is being processed at a time. We also sort the name servers
+      # to prevent potential deadlocks. For generating advisory lock key, we
+      # use crc32, as our usual choice, OpenSSL::HMAC requires base conversion
+      # and truncation whereas crc32 returns 32 bit integer, just want we want.
```

Comment:
```suggestion
      # and truncation whereas crc32 returns 32 bit integer, just what we want.
```

Reply:
Is it possible for someone to attack the interlock this way? crc32 is not so hard to attack pre-image attacked.

Also, what's wrong with multiple DNS actions at once, especially per-zone? Is bottlenecking all zones one on the other something we have to do?

Reply:
We use name server's id, which is randomly generated, so attack is quite difficult as attackers would not be able to decide that.

Multiple DNS actions per-zone is OK, I'll modify this to only lock based on zone.

Reply:
We don't do any locking thus not using crc32 anymore.